### ISSUE TITLE
Coverage: odd/non-square/anisotropic images, per-channel images, multi-channel adjoint vs ducc, channel-vmap numerics (T2)

### ISSUE THREAD

**Issue by chrisfinlay** [chrisfinlay/jax-nufft#14]
Part of https://github.com/chrisfinlay/jax-nufft/issues/2 · Tracking: #35 · Depends on: none

## Context
Verified correct in the review but never tested: odd sizes (33×33, 31×45), non-square (32×48, 48×32),
`pixsize_l ≠ pixsize_m`; per-channel distinct images (every multi-channel test broadcasts one 2-D image); the
multi-channel *adjoint* against ducc0 with weights; and `channel_strategy="vmap"` numerics
(`test_jit_static_strategy_args` only checks the output shape). ducc0's public API rejects odd sizes, so those
need the DFT reference.

## Implementation plan
1. `tests/test_against_dft.py`: parametrise the small forward/adjoint DFT tests over
   `image_shape, pixsize in [((32,32),(0.01,0.01)), ((33,33),(0.01,0.01)), ((32,48),(0.01,0.01)), ((31,45),(0.01,0.007)), ((32,48),(0.012,0.006))]`
   (300 rows, w ∈ ±40 m, eps = 1e-6) asserting `< 2 * eps`.
2. `tests/test_against_ducc.py::test_multichannel_adjoint_parity`: 4 channels over 20 % bandwidth, distinct vis and
   weights per channel; jax per-channel dirty images summed over channels equal ducc0's channel-summed image ≤ 3·eps
   (ducc0's `vis2dirty` sums channels); and each per-channel image against the DFT reference ≤ 2·eps.
3. `tests/test_against_ducc.py::test_multichannel_forward_per_channel_images`: 3-D image with different content per
   channel vs the DFT; both `channel_strategy` values; all four `w_strategy` values via parametrisation (fast: 48², 300 rows).
4. Extend `tests/test_jax_integration.py::test_jit_static_strategy_args` to compare `channel_strategy="vmap"` output
   against `"scan"` to 1e-11 (numeric, not just shape).
5. README API section: state that odd and non-square images are supported (ducc0 requires even).

## Definition of done
- [ ] New tests pass in the fast suite (< 10 s added).
- [ ] Any transpose/axis-order regression in `_dirty2vis_jit`/`_vis2dirty_jit` is caught by the non-square cases (verify by
      temporarily swapping `n_l`/`n_m` in `nufft1`'s output shape and confirming failure).

## Working context (self-contained)

- **Environment.** `pixi run -e test <cmd>` from the repo root (Python 3.12; jax, jax-finufft, numpy; ducc0 as the test
  oracle). Read `AGENTS.md` §1–§6 first: sign convention, plan-then-call pattern and the plan-field checklist (§4),
  strategies (§5), test/benchmark commands (§6). `pixi run -e test pytest -q --runslow`, `pixi run -e dev lint` and
  `pixi run -e dev typecheck` must stay green.
- **Fixtures.** `tests/conftest.py` defines telescopes `EDA2` (64², 120° FoV, 400 rows), `MWA_COMPACT` (128²),
  `MWA_EXTENDED` (256²), `MEERKAT` (256²), `GH200_LARGE` (2048², 50k rows) and `synthetic_uvw(tel, zenith_angle_deg, seed)`.
  "`<name>` zenith / off30" means `uvw = synthetic_uvw(tel, 0.0 or 30.0, seed=0)`, `freq = np.array([tel.freq_hz])`,
  `image_shape = (tel.n_pix, tel.n_pix)`, `pixsize = tel.pixsize`, image `np.random.default_rng(7).standard_normal(...)`
  and complex vis from the same generator. The "seven review fixtures" are EDA2 zenith, MWA_compact zenith and off30,
  MWA_extended zenith and off30, MeerKAT zenith and off30. The "five timing fixtures" are MWA_compact off30,
  MWA_extended off30, MeerKAT off30, EDA2 zenith, MWA_extended zenith.
- **Accuracy metric.** Relative L2 error `‖a − b‖₂ / ‖b‖₂` against the exact DFT in the repo's sign convention
  (`tests/test_against_dft.py::_reference_forward`, `tests/test_adjoint.py::_reference_adjoint`; vectorise over row
  chunks to make 256² fixtures take seconds). "x× eps" is that ratio divided by the requested epsilon. "The accuracy
  sweep" is the seven review fixtures × eps ∈ {1e-3, 1e-4, 1e-5, 1e-6, 1e-7, 1e-8, 1e-10, 1e-12}, forward and adjoint,
  `dense_scan` (the other strategies agree with it to 1e-14). The width-rule sub-issue adds
  `tests/test_accuracy_sweep.py` (gated by `--runsweep`) that produces this table; later issues re-run it.
- **ducc0 as oracle.** `ducc0.wgridder.dirty2vis(uvw=, freq=, dirty=, pixsize_x=, pixsize_y=, epsilon=, do_wgridding=True,
  divide_by_n=False, nthreads=1)` and `ducc0.wgridder.vis2dirty(..., npix_x=, npix_y=, divide_by_n=True)` match
  jax-nufft's default conventions (see `tests/test_against_ducc.py`). With `verbosity=1` ducc0 prints its plane count to
  C-level stdout as `(nu x nv x nplanes), supp=W`; capture it by redirecting file descriptor 1 (`os.dup2` to a temp file,
  not `contextlib.redirect_stdout`) and parse `r"\((\d+)x(\d+)x(\d+)\), supp=(\d+)"`.
- **Timing protocol.** Same machine before and after; `nthreads` stated; plan built and one warm-up call done outside the
  timer; median of ≥ 7 calls, each ending with `.block_until_ready()`. Timing gates are relative to the pre-PR number
  measured this way; absolute times quoted above are from the review machine (Apple M-series, 10 cores) for orientation only.
- **Memory protocol.** `jax.jit(fn).lower(*args).compile().memory_analysis().temp_size_in_bytes` (works on the CPU backend);
  compare with the same quantity for the plain forward/adjoint call.
- **GPU gates.** Need the CUDA jax-finufft build (pixi env `gpu-dev`, Linux). Without a GPU: satisfy the CPU gates, state in
  the PR that the GPU gate is pending for the maintainer, and use `docs/benchmarks/v0.1.2-baseline-gpu.json` (GH200
  per-cell median seconds, `n_w`, padding overhead) as the reference numbers.
- **PR conventions.** One PR per issue, branched from `dev`, title `<short> (closes #N)`, the definition-of-done checklist
  copied into the description with before/after numbers, and the statement that ducc0 source was not consulted.

Review finding IDs (private review document): T2, API5.

---
**Sources to implement from.** ducc0 is GPL-2.0-or-later; it is used in this project only as a black-box test
oracle through its public Python API. Do not consult or transliterate its source for this work; implement from the
mathematics stated above and these references (FINUFFT and jax-finufft are Apache-2.0 and may be read freely):

- [Arras+2021] P. Arras, M. Reinecke, R. Westermann, T. A. Enßlin, *Efficient wide-field radio interferometry response*, A&A 646, A58 (2021), arXiv:2010.10122 — §2 w-stacking, eq. 12 plane spacing, eq. 19 Hermitian symmetry, eqs. 20–21 plane placement, eq. 29 modified ES kernel, §3.2 + Appendix A optimal kernel parameters, §4.1 cost model.
- [Barnett+2019] A. H. Barnett, J. Magland, L. af Klinteberg, *A parallel nonuniform fast Fourier transform library based on an "exponential of semicircle" kernel*, SIAM J. Sci. Comput. 41(5), C479 (2019), arXiv:1808.06736 — eq. 8 ES kernel, eq. 10 β = 2.30·W and the width rule, Theorem 7 error decay, eq. 18 Gauss–Legendre kernel transform.
- [Ye+2022] H. Ye, S. F. Gull, S. M. Tan, B. Nikolic, *Optimal gridding and degridding in radio interferometry imaging*, MNRAS 510, 4110 (2022), arXiv:2110.03914 — kernel optimisation methodology.
- [TMS2017] A. R. Thompson, J. M. Moran, G. W. Swenson, *Interferometry and Synthesis in Radio Astronomy*, 3rd ed., Springer (2017), Ch. 3 — conjugate symmetry of the visibility function and the shift theorem.
- [Higham2002] N. J. Higham, *Accuracy and Stability of Numerical Algorithms*, 2nd ed., SIAM (2002), Ch. 1 — cancellation and range reduction.
- [FINUFFT] `src/spreadinterp.cpp::setup_spreader` (Apache-2.0) — the implemented width rule `W = ceil(-log10(eps/10))` at σ = 2.
- [jax-finufft] `lib/jax_finufft_cpu.cc` (Apache-2.0) — one `makeplan` per custom call, one `setpts` per point set.

**Evidence.** The measurements quoted above are from the 2026-09-02 review run (jax 0.9.2, jax-finufft 1.3.1,
ducc0 0.41.0 as oracle, macOS arm64 10-core, float64 unless stated). The review document is kept outside the
repository and is not to be committed; the definition-of-done tests below are the reproduction.

**Comment by chrisfinlay** [comment]
## A demonstrated defect this issue would have caught

Found while reviewing #20, which adds a `1/n` diagonal indexed on the image grid to **both**
operators.

Transposing that grid — a one-token change, `(plan.n_minus_1 + 1.0).astype(dtype).T` — passes the
**entire 1255-test suite**:

```
tests/test_divide_by_n.py   319 passed
full suite                 1255 passed, 530 skipped   (identical to baseline)
```

It is not a harmless mutation. On a plan with `pixsize_m = 0.6 · pixsize_l` (64², EDA2 zenith,
eps 1e-6, float64), `vis2dirty(divide_by_n=True)` against ducc0 at the same geometry goes from
**3.502e-07** — inside the 3·eps contract — to **1.022e+00**.

The reason nothing catches it is exactly this issue's subject: `plan.n_minus_1` is *exactly*
transpose-symmetric on every fixture in the repository, `max|nm1 − nm1.T| = 0.0` whenever
`pixsize_l == pixsize_m` and `n_l == n_m`. A repo-wide scan of every `make_plan` call site and shape
literal found **no non-square image and no anisotropic pixel size anywhere in `tests/`**. Any
quantity indexed on the image grid is therefore free to confuse its two axes.

#20 adds the minimum geometry needed to falsify its own diagonal — one anisotropic oracle cell and
one ducc0 parity cell — rather than widening into this issue's scope. What that leaves for here is
the general case: every other grid-indexed quantity in the codebase is under the same blind spot,
and a single anisotropic, non-square fixture in `conftest.py` would put all of them under test at
once.

Worth raising this issue's priority accordingly. It is currently effort S, and it is the only thing
standing between the repository and a class of axis-confusion bug that is invisible to 1255 tests
and catastrophic in output.

**Comment by chrisfinlay** [comment]
## Correction: my earlier comment overstated the blind spot

I wrote above that "every other grid-indexed quantity in the codebase is under the same blind
spot". Measured against current `main`, that is no longer true, and I am striking it.

#20 and #21 each added anisotropic fixtures that call the **whole operator** on an asymmetric plan
(`_ANISO_SQUARE` / `_ANISO_NONSQUARE` in `test_divide_by_n.py`, and #21's cotangent-indexing test).
Those incidentally cover every grid-indexed transpose in the forward, not just the `1/n` diagonal
they were written for. Transposing `phi_hat` on current `main` fails **six** pre-existing tests
before any of this issue's new cells are counted.

So the transposed-`1/n` demonstration in my earlier comment was real when measured, and the general
claim I drew from it has since been overtaken by the two PRs that followed.

## What was genuinely untested, and the defect it hides

**Odd extents. No odd image extent existed anywhere in `tests/`.**

`planning._n_minus_1_grid` computes `i = np.arange(n_l) - n_l // 2`. Changing the floor division to
a true division — `- n_l / 2`, which is what `README.md` writes as `l_i = (i - n_l/2)·pixsize_l` —
is identical for even extents and half a pixel off for odd ones. Under that change:

```
2 failed, 1478 passed, 546 skipped
FAILED tests/test_against_dft.py::test_geometry_matches_dft_forward_and_adjoint[odd_square]
FAILED tests/test_against_dft.py::test_geometry_matches_dft_forward_and_adjoint[odd_nonsquare_anisotropic]
```

Two failures in 1480 tests, both of them new cells from this issue. The shipped code is correct;
the README's formula is ambiguous about which convention it means, and nothing could tell them
apart.

## Four of this issue's premises are stale

Recorded so the next reader does not re-derive them:

1. **"every multi-channel test broadcasts one 2-D image"** — false since #23.
   `test_against_dft.py::test_multi_channel_matches_dft_forward_and_adjoint` already uses a 3-channel
   image with independent content per plane across all four `w_strategy` × both `channel_strategy`,
   with per-channel assertions; so do `test_strategies_equivalent.py` and `test_divide_by_n.py`. It
   is true only of `test_against_ducc.py`, which is where the new cell went.
2. **"`channel_strategy="vmap"` numerics (`test_jit_static_strategy_args` only checks the output
   shape)"** — the parenthetical is true, the headline is not.
   `test_strategies_equivalent.py::test_strategies_agree_pairwise` already compares all eight
   `(w_strategy, channel_strategy)` combinations pairwise at 1e-11 on three channels with distinct
   content. The change was made anyway because it strengthens the cell it lives in, but it is not
   the suite's first numeric vmap-vs-scan statement.
3. **The definition of done's mutation check is stale twice over.** "Swap `n_l`/`n_m` in `nufft1`'s
   output shape" does fail — but as a `jnp` broadcast error one line later (`mul got incompatible
   shapes for broadcasting: (48, 32), (32, 48)`), not at `_check_operand`, which validates the
   operand and never sees it. And swapping the primitive's declared output aval instead fails at
   MLIR lowering. Either way it is shape-only, so **only the non-square cells catch it** — which is
   why the square-anisotropic cell had to be added. It is not in this issue's list, and without it a
   quantity built with `pixsize_l`/`pixsize_m` exchanged, shape-valid on any grid, goes untested.
4. ducc0 0.41.0's refusal of odd sizes was confirmed empirically for both operators; non-square even
   shapes with anisotropic pixels do work. Both of this issue's assumptions there are correct.
